ENH: Timedelta isoformat

TomAugspurger opened the pull request:
Split from https://github.com/pandas-dev/pandas/pull/14904

A few questions before I finish all the docs / tests:

The [spec](https://en.wikipedia.org/wiki/ISO_8601#Durations) seems very permissive about what's allowed, so we need to choose

- Do we include components pandas can't represent like Years and Months with 0 `(P0Y0M...)` (no)
- Do we include components whose value is 0, e.g. `P0DT...` vs. `PT...` (yes)
- Do we 0 pad units, e.g. `...T5H` vs `T05H` (no)
- Do we strip trailing 0s in the seconds decimal: `...6.000000000S` vs. `6S` (yes)

We probably want a reader for this format as well, maybe in the Timedelta constructor.  Not sure if I'll have time to get to that before 0.20 . I'd rather prioritize the JSON table schema PR.

TomAugspurger committed:
ENH: Timedelta isoformat

jreback commented:
fyi the parsing issue is here: https://github.com/pandas-dev/pandas/issues/11375

jreback commented on pandas/tslib.pyx:
```diff
@@ -2966,6 +2966,49 @@ class Timedelta(_Timedelta):
+        """
+        Format Timedelta as ISO 8601 Duration like
+        `P[n]Y[n]M[n]DT[n]H[n]M[n]S`, where the `[n]`s are replaced by the
+        values for that duration.
```
maybe add the link to the 8601 format ?

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -91,6 +91,22 @@ support for bz2 compression in the python 2 c-engine improved (:issue:`14874`).
+`ISO8601 duration`_
+
+.. _ISO601 duration: https://en.wikipedia.org/wiki/ISO_8601
+
```
this link in the doc string as well

jreback commented:
lgtm

minor did comment

jreback commented on pandas/tseries/tests/test_timedeltas.py:
```diff
@@ -1347,6 +1347,35 @@ def test_components(self):
+        result = td.isoformat()
+        expected = 'P0DT0H0M0.000000123S'
+        self.assertEqual(result, expected)
+
```
add a test for NaT (which i think should just be nan), similar to what we do for Timestamp

TomAugspurger replied:
Good call. How about `NaT` to match `Timestamp`?

```python
In [24]: pd.Timestamp('NaT').isoformat()
Out[24]: 'NaT'

In [25]: pd.Timedelta('NaT').isoformat()
Out[25]: 'NaT'
```

jreback replied:
yep that's fine

TomAugspurger commented:
Thanks for the review. Added the `NaT` test and fixed up some doc issues. I also added a small example in `timedeltas.rst`, under the `Attributes` section. There wasn't really an appropriate place for it in `api.rst` that I saw.

jreback commented:
lgtm. merge on pass.

jorisvandenbossche commented:
Do we need to note that a `Timedelta` is not actually a 'duration' as defined by that standard? (as a timedelta specified absolute time difference and not a relative)

TomAugspurger commented:
@jorisvandenbossche does a `Timedelta` actual violate the definition? The way I'm reading it, a duration can mean either absolute or relative difference.
Something like `P1M`, for 1 month, is going to be relative and we can't represent that as a `Timedelta`. But the duration from '2016-01-01' to '2016-02-01' *can* be represented as a timedelta.

@pwalsh does that sound right to you?

jorisvandenbossche commented:
Yeah, it is not really clear from the wiki page. It depends on how the 'days' are interpreted, but from *"But keep in mind that "PT36H" is not the same as "P1DT12H" when switching from or to Daylight saving time."* I would conclude that the days are also relative and not absolute (not 1D == 24H), and I *think* for a timedelta this is absolute?

TomAugspurger commented:
@jorisvandenbossche rereading the wiki page again today. I missed this section earlier

> They should only be used as part of a time interval as prescribed by the standard. Time intervals are discussed in the next section
> [ ...]
> A time interval is the intervening time between two time points.

How do you read that? I'm going back and forth on it. I *think* that means we're OK, as python uses time deltas to represent the time between two points, and the spec allows

> Duration only, such as "P1Y2M10DT2H30M", with additional context information

As one valid representation. I am concerned that other libraries don't implement an `isoformat` on time deltas.

jreback commented:
``Timedelta`` is de-facto a duration, so I think the above definition is fine.

jorisvandenbossche commented:
> Timedelta is de-facto a duration,

Not in the sense that it is defined on the wiki (or at least the possible notations for a duration). For example, "1 month" is not representable by a Timedelta in general. 
But as Tom points out, they "should only be used as part of a time interval" according to the wiki, so which means that the duration is always interpreted based on a certain context. So in such a case, then also "1 month" is exactly defined in absolute time. Although the json table schema does not seem to follow that restriction of only using it as part of a time interval.

But anyway, this is probably too detailed discussion for this use case as long as we only specify days and below.

jreback commented:
@TomAugspurger my only remaining question is this in the JSON standard anywhere / common practice? (or is there *anything* for describing how to store durations / intervals)? 

For datetimes we have ISO and as epochs, is there such a thing for durations? (if so *maybe* we should have a kw arg ``timedelta_unit``?, though could be added later).

TomAugspurger commented:
- [This](https://www.npmjs.com/package/iso8601-duration) JS library converts absolute durations, like timedeltas, to iso8601 duration format.

- moment.js is quite clear that [its durations](https://momentjs.com/docs/#/durations/) are relative, not absolute

> Durations do not have a defined beginning and end date. They are contextless.
>
> A duration is conceptually more similar to '2 hours' than to 'between 2 and 4 pm today'. As such, they are not a good solution to converting between units that depend on context.

[They do convert to ISO8601](https://momentjs.com/docs/#/durations/as-json/)

- [this](https://pypi.org/project/isodate) python package seems to read and write between iso8601 durations and time deltas.

jreback commented:
ok @TomAugspurger then I think your ``.isoformat()`` is just fine, its a duration, and represents exactly what ``Timedelta`` is. (whether releative or absolute is context dependent, which is fine).

jreback commented:
merge on pass.

jreback commented:
thanks @TomAugspurger
